sql/sql_class.h

```diff
@@ -5169,6 +5170,23 @@ class THD : public MDL_context_owner,
   /// Flag indicating whether this session incremented the number of sessions
   /// with GTID_NEXT set to AUTOMATIC:tag
   bool has_incremented_gtid_automatic_count;
+
+ public:
+  bool register_kill_notification_hook(kill_notification_fn fn) {
```

Comment:
Do we need any protection from registering the same `fn` twice?

Reply:
Looks like most of my comments are for the first commit 🙂
In the second part you simply removed most of the suspicious places, which proves that it was a good idea to follow that road.

Reply:
Indeed, in the second commit the protection against this has been added.